### PR TITLE
Generate URL from route for the button_callback

### LINKED ISSUE #2951: No route support in DCA button_callbacks
**Affected version(s)**

4.7 or rather 4.9 (LTS)

**Description**

#116 introduced the 'route' parameter for DCA operations.

This URL is not passed to a custom button callback, e.g.

```php
TL_DCA.tl_talble.list.operations. =
            'my_oper' => [
                'route' => 'my_route',
            ],
```

```php
   /**
     * @Callback(table="tl_table", target="list.operations.my_oper.button")
     */
    public function attendancesButton(array $row, $href, $label, $title, $icon, $attributes)
    {
        // $href is empty but should be generated from route
    }
```

(I'll probably fix it by myself)

### PR DESCRIPTION
**Pull request by richardhj**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2951 
| Docs PR or issue | 

Generate the URL from the route parameter before passing it to the button_callback.

I decided to target this to 4.11 so you don't have to merge those PHP8 changes upstream 😅

### CODE REVIEW

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 1]
```diff
@@ -797,17 +797,23 @@ protected function generateButtons($arrRow, $strTable, $arrRootIds=array(), $bln
...
+			{
+				$href = System::getContainer()->get('router')->generate($v['route'], array('id' => $arrRow['id']));
+			}
+
 			// Call a custom function instead of using the default button
 			if (\is_array($v['button_callback'] ?? null))
 			{
 				$this->import($v['button_callback'][0]);
-				$return .= $this->{$v['button_callback'][0]}->{$v['button_callback'][1]}($arrRow, $v['href'] ?? null, $label, $title, $v['icon'] ?? null, $attributes, $strTable, $arrRootIds, $arrChildRecordIds, $blnCircularReference, $strPrevious, $strNext, $this);
+				$return .= $this->{$v['button_callback'][0]}->{$v['button_callback'][1]}($arrRow, $href, $label, $title, $v['icon'] ?? null, $attributes, $strTable, $arrRootIds, $arrChildRecordIds, $blnCircularReference, $strPrevious, $strNext, $this);
```
Comment (m-vo) [inline core-bundle/src/Resources/contao/classes/DataContainer.php:811]: The 2nd parameter of the `button_callback` (href) will now never be `null`, right? 

Are there cases we would want to transform this (`$href ?: null`)? If not we should probably adjust the [documentation](https://docs.contao.org/dev/reference/dca/callbacks/#list-global-operations-operation-button).

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 2]
```diff
@@ -919,25 +925,27 @@ protected function generateGlobalButtons()
 				$title = $label;
 			}
 
+			$href = $v['href'] ?? '';
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
			$href = $v['href'] ?? null;
```

**Review by aschempp — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/classes/DataContainer.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/DataContainer.php:811]: Fixed in 45758bc2d3e70cde3dc26c4883220424715e7f29.

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 3]
```diff
@@ -919,25 +926,28 @@ protected function generateGlobalButtons()
...
 			}
 
-			if (!empty($v['route']))
-			{
-				$href = System::getContainer()->get('router')->generate($v['route']);
-			}
-			else
+			if (empty($v['route']))
 			{
 				$href = $this->addToUrl($v['href']);
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
				$href = $this->addToUrl($v['href'] ?? null);
```

Should this not be `$v['href'] ?? null`?

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 4]
```diff
@@ -1042,7 +1052,7 @@ protected function generateHeaderButtons($arrRow, $strPtable)
 			{
 				if (!empty($v['route']))
 				{
-					$href = System::getContainer()->get('router')->generate($v['route'], array('id' => $arrRow['id']));
+					$href = System::getContainer()->get('router')->generate($v['route'], array('id' => $arrRow['id'], 'nc' => Input::get('nb') ? '1' : null));
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/DataContainer.php:1055]: Why is this change necessary?

**Review by leofeyer — CHANGES_REQUESTED** [review]
You are still using `$v['href']` instead of `$href` in lines 825 and 838. Is this intentional or did you overlook the occurrences?

**Comment by fritzmg** [comment]
Currently, if you use a `route` or no `href` or `route` option at all for your `global_operation` (because you generate the appropriate url within the callback) the following error will appear in PHP 8:

```
ErrorException:
Warning: Undefined array key "href"

  at vendor\contao\contao\core-bundle\src\Resources\contao\classes\DataContainer.php:935
  at Contao\DataContainer->generateGlobalButtons()
     (vendor\contao\contao\core-bundle\src\Resources\contao\drivers\DC_Table.php:4822)
  at Contao\DC_Table->listView()
     (vendor\contao\contao\core-bundle\src\Resources\contao\drivers\DC_Table.php:350)
  at Contao\DC_Table->showAll()
     (vendor\contao\contao\core-bundle\src\Resources\contao\classes\Backend.php:648)
  at Contao\Backend->getBackendModule('news', null)
     (vendor\contao\contao\core-bundle\src\Resources\contao\controllers\BackendMain.php:167)
  at Contao\BackendMain->run()
     (vendor\contao\contao\core-bundle\src\Controller\BackendController.php:49)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor\symfony\http-kernel\HttpKernel.php:156)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor\symfony\http-kernel\HttpKernel.php:78)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor\symfony\http-kernel\Kernel.php:199)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public\index.php:31)
  at require('public/index.php')
     (public\app.php:4)            
```

This PR would fix this as far as I can see. @richardhj can you finish this? :)
